x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/PolicyStepsRegistry.java

```diff
@@ -75,41 +112,85 @@ public void update(IndexLifecycleMetadata meta, Client client, LongSupplier nowS
                 List<Step> policyAsSteps = policyMetadata.getPolicy().toSteps(policyClient, nowSupplier);
                 if (policyAsSteps.isEmpty() == false) {
                     firstStepMap.put(policyMetadata.getName(), policyAsSteps.get(0));
-                    stepMap.put(policyMetadata.getName(), new HashMap<>());
-                    Map<Step.StepKey, Step> stepMapForPolicy = stepMap.get(policyMetadata.getName());
+                    final Map<Step.StepKey, Step> stepMapForPolicy = new HashMap<>();
                     for (Step step : policyAsSteps) {
-                        assert ErrorStep.NAME.equals(step.getKey().getName()) == false;
+                        assert ErrorStep.NAME.equals(step.getKey().getName()) == false : "unexpected error step in policy";
                         stepMapForPolicy.put(step.getKey(), step);
                     }
+                    stepMap.put(policyMetadata.getName(), stepMapForPolicy);
                 }
             }
         }
 
-        for (String deletedPolicyName : mapDiff.getDeletes()) {
-            lifecyclePolicyMap.remove(deletedPolicyName);
-            firstStepMap.remove(deletedPolicyName);
-            stepMap.remove(deletedPolicyName);
+        for (ObjectCursor<IndexMetaData> imd : clusterState.metaData().getIndices().values()) {
+            final String indexName = imd.value.getIndex().getName();
+            final String policy = imd.value.getSettings().get(LifecycleSettings.LIFECYCLE_NAME);
+            if (policy == null) {
+                indexPhaseSteps.remove(indexName);
+            } else {
+                final List<Step> currentSteps = indexPhaseSteps.get(indexName);
+                // Get the current steps' phase, if there are steps stored
+                final String existingPhase = (currentSteps == null || currentSteps.size() == 0) ?
+                    "_none_" : currentSteps.get(0).getKey().getPhase();
+                // Retrieve the current phase, defaulting to "new" if no phase is set
+                final String currentPhase = imd.value.getSettings().get(LifecycleSettings.LIFECYCLE_PHASE, "new");
+                assert currentPhase != null : "expected a current phase but it was null";
```

Comment:
is this necessary given that the Settings.get with default will never have this be `null`?

Reply:
I think this was a holdover from before I added the default return value for the setting retrieval, I'll remove it